docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -0,0 +1,403 @@
+
+As most sources are used by all taint tracking queries there are only a few different source kinds.
+The following source kinds are supported:
+
+- **remote**: A remote source of possibly tainted data. This is the most common kind for a source. Sources of this kind is used for almost all taint tracking queries.
+- **contentprovider**: ?
+- **android-widget**: ?
+- **android-external-storage-dir**: ?
+
+sinkModel(package, type, subtypes, name, signature, ext, input, kind, provenance)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Taint sink. As opposed to source kinds, there are many different kinds of sinks as these tend to be more query specific.
+
+- **input**: Access path to the sink, where we want to check if tainted data can flow to.
+- **kind**: Kind of the sink.
+
+The following sink kinds are supported:
+
+- **open-url**: ?
+- **jndi-injection**: ?
+- **ldap**: ?
+- **sql**: ?
+- **jdbc-url**: ?
+- **logging**: ?
+- **mvel**: ?
+- **xpath**: ?
+- **groovy**: ?
+- **xss**: ?
+- **ognl-injection**: ?
+- **intent-start**: ?
+- **pending-intent-sent**: ?
+- **url-open-stream**: ?
+- **url-redirect**: ?
+- **create-file**: ?
+- **read-file**: ?
+- **write-file**: ?
+- **set-hostname-verifier**: ?
+- **header-splitting**: ?
+- **information-leak**: ?
+- **xslt**: ?
+- **jexl**: ?
+- **bean-validation**: ?
+- **ssti**: ?
+- **fragment-injection**: ?
+
+summaryModel(package, type, subtypes, name, signature, ext, input, output, kind, provenance)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Flow through. This extension point is used to model flow through elements.
+
+- **input**: Access path to the input of the element (where data will flow to the output).
+- **output**: Access path to the output of the element (where data will flow from the input).
+- **kind**: Kind of the flow through.
+- **provenance**: Provenance (origin) of the flow through.
+
+The following kinds are supported:
+
+- **taint**: This means the output is not necessarily equal to the input, but it was derived from the input in an unrestrictive way. An attacker who controls the input will have significant control over the output as well.
+- **value**: This means that the output equals the input or a copy of the input such that all of its properties are preserved.
+
+neutralModel(package, type, name, signature, provenance)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Access paths
+------------
+The **input**, and **output** columns consist of a **.**-separated list of components, which is evaluted from left to right, with each step selecting a new set of values derived from the previous set of values.
+
+The following components are supported:
+
+- **Argument[**\ `n`\ **]** selects the argument at index `n` (zero-indexed).
+- **Argument[**\ `-1`\ **]** selects the qualifier.
+- **Argument[**\ `n1..n2`\ **]** selects the arguments in the given range (both ends included).
```

Comment:
How do you specify all arguments? What about varargs?

Reply:
I don't think that has been implemented for Java. You need to be explicit. Some languages has implemented support for eg. `Argument[any]` or `Argument[0..]` (open ended intervals), but AFAIK this is not the case for Java (there are quite a lot of discrepancies between the modelling capabilities of the different languages).

Reply:
Thanks for the explanation. I can see these subtleties as a source for confusion. When we roll out docs for other languages, it would be nice to have a location where we document all the discrepancies between the languages.

Reply:
Yes, that is an interesting idea, but I am not sure whether that is important for external users.
I would imagine that if you are a security researcher (or similar persona) in an organisation then most likely you are only working with one - maybe two languages (I have no information to back up this claim - do you know anything about this @coadaflorin ?).
However, for internal purposes it would be interesting to uncover these differences as these might simply be development gaps and ideas that were not ported between the languages.

Reply:
FWIW, varargs are extracted as a single array argument, so you can select them by using the syntax `Argument[n]` where `n` is the last argument.

Reply:
Ups yes - I didn't answer that part! Thank you @atorralba !

Reply:
> internal purposes it would be interesting to uncover these differences as these might simply be development gaps and ideas that were not ported between the languages.

I think that's fine. These are not meant to be publicly facing docs (even though they are in a public repo), so we don't require the same level of completeness for them. Though, these differences are something that we should be aware of, especially when we do a public release.